crates/data_components/src/databricks/sql_warehouse.rs

```diff
@@ -0,0 +1,655 @@
+                reason: format!("data_array[{i}] lacks col_name or data_type"),
+            });
+        }
+
+        let col_name = row_array[0]
+            .as_str()
+            .ok_or_else(|| Error::UnableToRetrieveSchema {
+                reason: format!("data_array[{i}][0] is not a string"),
+            })?;
+
+        // If we see #, DB is now providing the clustering metadata info and
+        // we have all the columns we need
+        if col_name.starts_with('#') {
+            break;
+        }
+
+        let data_type_str = row_array[1]
+            .as_str()
+            .ok_or_else(|| Error::UnableToRetrieveSchema {
+                reason: format!("data_array[{i}][1] is not a string"),
+            })?;
+
+        let field = Field::new(col_name, map_databricks_type(data_type_str)?, true);
+
+        fields.push(field);
+    }
+
+    Ok(Arc::new(Schema::new(fields)))
+}
+
+struct SqlWarehouseConnection {
+    api: Arc<SqlWarehouseApi>,
+}
+
+impl<'a> DbConnection<Arc<SqlWarehouseApi>, &'a dyn Sync> for SqlWarehouseConnection {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+
+    fn as_async(&self) -> Option<&dyn AsyncDbConnection<Arc<SqlWarehouseApi>, &'a dyn Sync>> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl<'a> AsyncDbConnection<Arc<SqlWarehouseApi>, &'a dyn Sync> for SqlWarehouseConnection {
+    fn new(api: Arc<SqlWarehouseApi>) -> Self {
+        Self { api }
+    }
+
+    async fn get_schema(
+        &self,
+        table_reference: &TableReference,
+    ) -> Result<SchemaRef, dbconnection::Error> {
+        self.api
+            .get_schema(table_reference)
+            .await
+            .map_err(|source| dbconnection::Error::UnableToGetSchema {
+                source: Box::new(source),
+            })
+    }
+
+    async fn query_arrow(
+        &self,
+        sql: &str,
+        _: &[&'a dyn Sync],
+        _projected_schema: Option<SchemaRef>,
+    ) -> Result<SendableRecordBatchStream, Box<dyn std::error::Error + Send + Sync>> {
+        let token = self.api.token_provider.get_token();
+        let payload = json!({
+            "warehouse_id": self.api.sql_warehouse_id,
+            "format": "ARROW_STREAM",
+            "disposition": "EXTERNAL_LINKS",
+            "statement": sql,
+        });
+
+        let mut response = self.api.execute_request(&token, &payload).await?;
+
+        // Get the result object
+        let result_object = response
+            .get_mut("result")
+            .map(Value::take)
+            .ok_or_else(|| MissingJsonFieldSnafu { field: "result" }.build())?;
+
+        Ok(SqlWarehouseApi::fetch_external_links(Arc::clone(&self.api), result_object).await?)
+    }
+
+    async fn execute(
+        &self,
+        _query: &str,
+        _: &[&'a dyn Sync],
+    ) -> Result<u64, Box<dyn std::error::Error + Send + Sync>> {
+        Ok(NotImplementedSnafu.fail()?)
```

Comment:
Is this expected?

Reply:
I was using the snowflake implementation as a reference which also leaves this API unimplemented. We can implement this if we want write operations in a follow-up.